pandas/core/indexing.py

```diff
@@ -630,6 +630,17 @@ def can_do_equal_len():
                     self.obj[item_labels[indexer[info_axis]]] = value
                     return
 
+                if isinstance(value, ABCDataFrame):
```

Comment:
this is very complicated and likely does not belong here. you would need to heavily comment this and explain line by line.